// ng-location-picker/index.js
import { Directive, EventEmitter, HostListener, Input, NgModule, Output, ViewChild } from '@angular/core';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var HTMLFactory = (function () {
    function HTMLFactory() {
    }
    /**
     * @param {...?} children
     * @return {?}
     */
    HTMLFactory.createLocationSelectorElement = /**
     * @param {...?} children
     * @return {?}
     */
    function () {
        var children = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            children[_i] = arguments[_i];
        }
        var /** @type {?} */ locationSelectorElement = HTMLFactory.createElement('div', 'location-selector');
        for (var /** @type {?} */ i = 0; i < children.length; i++) {
            var /** @type {?} */ child = children[i];
            locationSelectorElement.appendChild(child);
        }
        document.body.appendChild(locationSelectorElement);
        return locationSelectorElement;
    };
    /**
     * @param {?} label
     * @param {?} clickAction
     * @return {?}
     */
    HTMLFactory.createDoneButton = /**
     * @param {?} label
     * @param {?} clickAction
     * @return {?}
     */
    function (label, clickAction) {
        var /** @type {?} */ doneButtonElement = HTMLFactory.createElement('button', 'done-button', 'control done-button');
        doneButtonElement.addEventListener('click', clickAction);
        doneButtonElement.innerText = label;
        return doneButtonElement;
    };
    /**
     * @return {?}
     */
    HTMLFactory.createStyles = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ styles = document.createElement('style');
        styles.innerText = '#location-selector {position: fixed !important;top: 0;bottom: 0;left: 0;right: 0;z-index: 2;text-align: center;overflow: hidden;} ' +
            '#location-selector .map {position: absolute;top: 0;bottom: 0;left: 0;right: 0;} ' +
            '#location-selector .control {overflow: hidden;position: absolute;color: #fff;font-family: Roboto, Arial, sans-serif;font-size: 11px;background-color: #000;padding: 11px;box-shadow: rgba(0, 0, 0, 0.3) 0px 1px 4px -1px;border: none;border-radius: 4px}' +
            '#location-selector .search-box {top: 11px;z-index: 3;width: 200px;left: calc(50% - 100px)}' +
            '#location-selector .search-box::placeholder {color: #fff}' +
            '#location-selector .done-button {bottom: 15px;z-index: 3;width: 50px;left: calc(50% - 25px)}';
        return styles;
    };
    /**
     * @return {?}
     */
    HTMLFactory.createSearchBoxElement = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ searchBoxElement = HTMLFactory.createElement('input', 'location-selector-search-box', 'control search-box');
        searchBoxElement.setAttribute('placeholder', 'Search');
        return searchBoxElement;
    };
    /**
     * @return {?}
     */
    HTMLFactory.createMapElement = /**
     * @return {?}
     */
    function () {
        return HTMLFactory.createElement('div', 'location-selector-map', 'map');
    };
    /**
     * @param {?} elementType
     * @param {?} id
     * @param {?=} styleClasses
     * @return {?}
     */
    HTMLFactory.createElement = /**
     * @param {?} elementType
     * @param {?} id
     * @param {?=} styleClasses
     * @return {?}
     */
    function (elementType, id, styleClasses) {
        var /** @type {?} */ locationSelectorElement = document.createElement(elementType);
        locationSelectorElement.setAttribute('id', id);
        locationSelectorElement.setAttribute('class', styleClasses ? styleClasses : '');
        return locationSelectorElement;
    };
    return HTMLFactory;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var MapStyles = (function () {
    function MapStyles() {
    }
    MapStyles.clearStyle = [
        {
            'featureType': 'poi',
            'elementType': 'labels.text',
            'stylers': [
                {
                    'visibility': 'off'
                }
            ]
        },
        {
            'featureType': 'poi.business',
            'stylers': [
                {
                    'visibility': 'off'
                }
            ]
        },
        {
            'featureType': 'road',
            'elementType': 'labels.icon',
            'stylers': [
                {
                    'visibility': 'off'
                }
            ]
        },
        {
            'featureType': 'transit',
            'stylers': [
                {
                    'visibility': 'off'
                }
            ]
        }
    ];
    return MapStyles;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var LatLng = google.maps.LatLng;
var Map = google.maps.Map;
var MapCreator = (function () {
    function MapCreator() {
    }
    /**
     * @param {?} mapElement
     * @param {?=} center
     * @return {?}
     */
    MapCreator.createMap = /**
     * @param {?} mapElement
     * @param {?=} center
     * @return {?}
     */
    function (mapElement, center) {
        var /** @type {?} */ mapOptions = {
            center: center ? center : new LatLng(4.6361512, -74.0832687),
            zoom: 15,
            mapTypeId: google.maps.MapTypeId.ROADMAP,
            mapTypeControl: false,
            fullscreenControl: false,
            styles: MapStyles.clearStyle
        };
        return new Map(mapElement, mapOptions);
    };
    return MapCreator;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var Marker = google.maps.Marker;
var MarkerService = (function () {
    function MarkerService() {
    }
    /**
     * @return {?}
     */
    MarkerService.clear = /**
     * @return {?}
     */
    function () {
        MarkerService.marker = null;
    };
    /**
     * @param {?} place
     * @param {?} map
     * @return {?}
     */
    MarkerService.showPlaceInMap = /**
     * @param {?} place
     * @param {?} map
     * @return {?}
     */
    function (place, map) {
        MarkerService.showLocationInMap(map, place.geometry.location, place.icon, place.name);
        map.setZoom(16);
    };
    /**
     * @param {?} map
     * @param {?} location
     * @param {?=} iconUrl
     * @param {?=} title
     * @return {?}
     */
    MarkerService.showLocationInMap = /**
     * @param {?} map
     * @param {?} location
     * @param {?=} iconUrl
     * @param {?=} title
     * @return {?}
     */
    function (map, location, iconUrl, title) {
        if (MarkerService.marker) {
            MarkerService.marker.setMap(null);
        }
        MarkerService.marker = MarkerService.createMarker(map, location, iconUrl, title);
        map.panTo(location);
    };
    /**
     * @param {?} map
     * @param {?} position
     * @param {?=} iconUrl
     * @param {?=} title
     * @return {?}
     */
    MarkerService.createMarker = /**
     * @param {?} map
     * @param {?} position
     * @param {?=} iconUrl
     * @param {?=} title
     * @return {?}
     */
    function (map, position, iconUrl, title) {
        var /** @type {?} */ icon = iconUrl ? this.createIcon(iconUrl) : this.createIcon('assets/icons/location.svg');
        return new Marker({
            map: map,
            title: title,
            animation: google.maps.Animation.DROP,
            position: position,
            icon: icon
        });
    };
    /**
     * @param {?} iconUrl
     * @return {?}
     */
    MarkerService.createIcon = /**
     * @param {?} iconUrl
     * @return {?}
     */
    function (iconUrl) {
        return {
            url: iconUrl,
            origin: new google.maps.Point(0, 0),
            scaledSize: new google.maps.Size(50, 50)
        };
    };
    MarkerService.marker = null;
    return MarkerService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var SearchBox = google.maps.places.SearchBox;
var LatLng$1 = google.maps.LatLng;
var LocationPickerDirective = (function () {
    function LocationPickerDirective() {
        this.onPickedLocation = new EventEmitter();
    }
    /**
     * @return {?}
     */
    LocationPickerDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this.pickedLocationIcon) {
            this.pickedLocationIcon = 'data:image/svg+xml;base64,PHN2ZyBmaWxsPSIjMDAwMDAwIiBoZWlnaHQ9IjQ4IiB2aWV3Qm94PSIwIDAgMjQgMjQi' +
                'IHdpZHRoPSI0OCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KICAgIDxwYXRoIGQ9Ik0xMiAyQzguMTMgMiA1IDUuMTMgNSA5YzAgNS' +
                '4yNSA3IDEzIDcgMTNzNy03Ljc1IDctMTNjMC0zLjg3LTMuMTMtNy03LTd6bTAgOS41Yy0xLjM4IDAtMi41LTEuMTItMi41LTIuNXMxLjEyLTIuNSAyLjUt' +
                'Mi41IDIuNSAxLjEyIDIuNSAyLjUtMS4xMiAyLjUtMi41IDIuNXoiLz4KICAgIDxwYXRoIGQ9Ik0wIDBoMjR2MjRIMHoiIGZpbGw9Im5vbmUiLz4KPC9zdmc+';
        }
    };
    /**
     * @return {?}
     */
    LocationPickerDirective.prototype.onClick = /**
     * @return {?}
     */
    function () {
        this.createHTML();
        this.createMap();
        this.centerMapOnGivenLocation();
        this.createSearchBox();
        this.enableSelectPosition();
    };
    /**
     * @return {?}
     */
    LocationPickerDirective.prototype.createHTML = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var /** @type {?} */ styles = HTMLFactory.createStyles();
        var /** @type {?} */ doneButton = HTMLFactory.createDoneButton('Done', function () {
            document.body.removeChild(_this.locationSelectorElement);
            _this.mapElement = null;
            _this.locationSelectorElement = null;
            _this.searchBoxElement = null;
        });
        this.searchBoxElement = HTMLFactory.createSearchBoxElement();
        this.mapElement = HTMLFactory.createMapElement();
        this.locationSelectorElement = HTMLFactory.createLocationSelectorElement(styles, doneButton, this.searchBoxElement, this.mapElement);
    };
    /**
     * @return {?}
     */
    LocationPickerDirective.prototype.createMap = /**
     * @return {?}
     */
    function () {
        this.mapElement = document.getElementById('location-selector-map');
        this.map = MapCreator.createMap(this.mapElement);
    };
    /**
     * @return {?}
     */
    LocationPickerDirective.prototype.centerMapOnGivenLocation = /**
     * @return {?}
     */
    function () {
        if (this.defaultCenter) {
            this.map.setCenter(this.defaultCenter);
        }
    };
    /**
     * @return {?}
     */
    LocationPickerDirective.prototype.enableSelectPosition = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.map.addListener('click', function (event) {
            MarkerService.showLocationInMap(_this.map, event.latLng, _this.pickedLocationIcon);
            _this.savePickedLocation(event.latLng);
        });
    };
    /**
     * @return {?}
     */
    LocationPickerDirective.prototype.createSearchBox = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.searchBox = new SearchBox(this.searchBoxElement);
        this.searchBoxElement.addEventListener('click', function (event) {
            event.target['value'] = '';
        });
        this.map.addListener('bounds_changed', function () {
            _this.searchBox.setBounds(_this.map.getBounds());
        });
        this.searchBox.addListener('places_changed', function () {
            var /** @type {?} */ places = _this.searchBox.getPlaces();
            if (places.length > 0 && places[0].geometry) {
                MarkerService.showPlaceInMap(places[0], _this.map);
                _this.savePickedLocation(places[0].geometry.location);
            }
        });
    };
    /**
     * @param {?} location
     * @return {?}
     */
    LocationPickerDirective.prototype.savePickedLocation = /**
     * @param {?} location
     * @return {?}
     */
    function (location) {
        if (this.onPickedLocation) {
            this.onPickedLocation.emit(location.toJSON());
        }
    };
    LocationPickerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ng-location-picker]'
                },] },
    ];
    /** @nocollapse */
    LocationPickerDirective.ctorParameters = function () { return []; };
    LocationPickerDirective.propDecorators = {
        "pickedLocationIcon": [{ type: Input },],
        "defaultCenter": [{ type: Input },],
        "onPickedLocation": [{ type: Output },],
        "searchBoxElement": [{ type: ViewChild, args: ['searchBox',] },],
        "onClick": [{ type: HostListener, args: ['click',] },],
    };
    return LocationPickerDirective;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var LocationPickerModule = (function () {
    function LocationPickerModule() {
    }
    /**
     * @return {?}
     */
    LocationPickerModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: LocationPickerModule,
            providers: []
        };
    };
    LocationPickerModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [LocationPickerDirective],
                    exports: [LocationPickerDirective]
                },] },
    ];
    /** @nocollapse */
    LocationPickerModule.ctorParameters = function () { return []; };
    return LocationPickerModule;
}());

export { LocationPickerModule, LocationPickerDirective };
